src/main/java/frc/robot/swerve/ModuleIOReal.java

```diff
@@ -0,0 +1,181 @@
+package frc.robot.swerve;
+
+import com.ctre.phoenix6.configs.Slot0Configs;
+import com.ctre.phoenix6.configs.TalonFXConfiguration;
+import com.ctre.phoenix6.controls.*;
+import com.ctre.phoenix6.hardware.TalonFX;
+import com.ctre.phoenix6.signals.NeutralModeValue;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.kinematics.SwerveModulePosition;
+import edu.wpi.first.math.util.Units;
+import edu.wpi.first.wpilibj.DutyCycleEncoder;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.RunCommand;
+import utils.math.AngleUtil;
+import utils.math.differential.Integral;
+
+public class ModuleIOReal implements ModuleIO {
+
+    private final TalonFX driveMotor;
+    private final TalonFX angleMotor;
+    private final DutyCycleEncoder encoder;
+    private final Integral driveSupplyChargeUsedCoulomb = new Integral();
+    private final Integral driveStatorChargeUsedCoulomb = new Integral();
+    private final Integral angleSupplyChargeUsedCoulomb = new Integral();
+    private final Integral angleStatorChargeUsedCoulomb = new Integral();
+    private double angleSetpoint;
+    private double currentAngle;
+    private double driveMotorVelocitySetpoint;
+
+    public ModuleIOReal(int driveMotorID, int angleMotorID, int encoderID,
+                        double[] motionMagicConfigs, int number) {
+        this.driveMotor = new TalonFX(driveMotorID);
+        this.angleMotor = new TalonFX(angleMotorID);
+
+        this.encoder = new DutyCycleEncoder(encoderID);
+
+        TalonFXConfiguration driveMotorConfig = new TalonFXConfiguration();
+        Slot0Configs drivePIDGains = new Slot0Configs()
+                .withKP(SwerveConstants.DRIVE_kP)
+                .withKI(SwerveConstants.DRIVE_kI)
+                .withKD(SwerveConstants.DRIVE_kD)
+                .withKV(SwerveConstants.DRIVE_KF);
+
+        driveMotorConfig.Slot0 = drivePIDGains;
+        driveMotorConfig.Voltage.PeakForwardVoltage = SwerveConstants.VOLT_COMP_SATURATION;
+        driveMotorConfig.Voltage.PeakReverseVoltage = -SwerveConstants.VOLT_COMP_SATURATION;
+        driveMotorConfig.CurrentLimits.SupplyCurrentLimit = SwerveConstants.SUPPLY_CURRENT_LIMIT.currentLimit;
+        driveMotorConfig.CurrentLimits.StatorCurrentLimit = SwerveConstants.STATOR_CURRENT_LIMIT.currentLimit;
+        driveMotorConfig.ClosedLoopGeneral.ContinuousWrap = true;
+        driveMotorConfig.Feedback.SensorToMechanismRatio = 1 / SwerveConstants.DRIVE_REDUCTION;
+        driveMotorConfig.Feedback.RotorToSensorRatio = 1;
```

Comment:
All that should be in the constants file.

Reply:
I think it will make the constants file messy and make it harder to check if a motor has all the configs it should.

Reply:
On the contrary, right now you have settings both here and in the constants file. Having everything in the same place will make finding settings easier.

Anyway, with how it's written now, you need to jump to the constants file anyway -- for example, in https://github.com/Galaxia5987/common/blob/5afeda19ac2330d7fb26c43a38171469a48009e7/src/main/java/frc/robot/swerve/ModuleIOReal.java#L48, you can't check the value directly. It's better to declare the configuration object directly in the constants file to see everything in one glance (also use the `.with...()` config methods, they help decrease the code length).

Reply:
ca87e4b2668351dd858d8a56fdf9504c2622d339